qiskit/qpy/binary_io/circuits.py

```diff
@@ -355,27 +354,28 @@ def _read_custom_instructions(file_obj, version, vectors):
     return custom_instructions
 
 
-def _write_instruction_parameter(file_obj, param):
+def _dumps_instruction_parameter(param):
     if isinstance(param, QuantumCircuit):
-        type_key = common.ProgramTypeKey.CIRCUIT
-        data = common.data_to_binary(param, write_circuit)
+        type_key = type_keys.Program.CIRCUIT
+        data_bytes = common.data_to_binary(param, write_circuit)
     elif isinstance(param, range):
-        type_key = common.ContainerTypeKey.RANGE
-        data = struct.pack(formats.RANGE_PACK, param.start, param.stop, param.step)
+        type_key = type_keys.Container.RANGE
+        data_bytes = struct.pack(formats.RANGE_PACK, param.start, param.stop, param.step)
     elif isinstance(param, tuple):
-        type_key = common.ContainerTypeKey.TUPLE
-        data = common.sequence_to_binary(param, _write_instruction_parameter)
+        type_key = type_keys.Container.TUPLE
+        data_bytes = common.sequence_to_binary(param, _dumps_instruction_parameter)
     elif isinstance(param, int):
         # TODO This uses little endian. This should be fixed in next QPY version.
-        type_key = common.ValueTypeKey.INTEGER
-        data = struct.pack("<q", param)
+        type_key = type_keys.Value.INTEGER
+        data_bytes = struct.pack("<q", param)
     elif isinstance(param, float):
         # TODO This uses little endian. This should be fixed in next QPY version.
-        type_key = common.ValueTypeKey.FLOAT
-        data = struct.pack("<d", param)
+        type_key = type_keys.Value.FLOAT
+        data_bytes = struct.pack("<d", param)
     else:
-        type_key, data = value.dumps_value(param)
-    common.write_instruction_param(file_obj, type_key, data)
+        type_key, data_bytes = value.dumps_value(param)
+
+    return type_key, data_bytes
```

Comment:
for review; this change is made to make sure sequence write uses `INSTRUCTION_PARAM` struct as stated in the QPY API docs.